### PR TITLE
Add chart for Hadoop

### PR DESCRIPTION
**Pull request by danisla**
Chart for running Hadoop and YARN as StatefulSets. Based on initial K8S work [here](github.com/comcast/kube-yarn).

### CODE REVIEW

**Comment by unguiculus** [comment]
Thanks for the PR. I hope you'll get the CLA sorted.

Please have a look at our best practices, especially regarding standard labels.

https://github.com/kubernetes/helm/blob/master/docs/chart_best_practices/labels.md

Also, see chart examples for reference:

https://github.com/kubernetes/helm/tree/master/docs/examples

BTW, creating a new chart with helm create Helm 2.5+ will give you an initial experience with best practices applied.

**Comment by danisla** [comment]
Ok I've update the chart with the standard labels.

**Comment by viglesiasce** [comment]
@danisla can you make this chart support 1.6 and 1.7, we strive to support the latest minor and 1 back.

The error is:
```
W0818 22:40:39.951] Error: error validating "": error validating data: found invalid field podManagementPolicy for v1beta1.StatefulSetSpec
```

**Comment by danisla** [comment]
Got it, should be compatible now.

/retest

**Comment by unguiculus** [comment]
Please check the chart examples again, especially regarding the `app` label and the `_helpers.tpl`.

**Comment by danisla** [comment]
Am I not allowed to add a postfix to the app label like it did for each component? Does it have to be exact per the generated output of `helm create`? Should I add a new label, `component` to do achieve the same thing?

I updated the `_helpers.tpl`, it was using an older version. You should consider make this an automated check as part of the jenkins build.

**Comment by unguiculus** [comment]
`app` labels should be `{{ template "name" . }}` on all resources without any suffix. You may append component suffixes to resources names if that makes sense. It would also be good to add ´component´ labels in such cases. Label selectors should have `app`, `release`, and, if applicable, `component` labels, but not `chart` and `heritage` labels.

**Comment by unguiculus** [comment]
Marking this as stale. Please update within one week.

**Comment by danisla** [comment]
Ok I have updated the labels, look good @unguiculus?

**Comment by unguiculus** [comment]
Please read my comment above again and checkout the chart examples at https://github.com/kubernetes/helm/tree/master/docs/examples/nginx. FYI, if you use `helm create` with Helm 2.5+, you get correct labels out of the box.

Also, we've started namespacing templates. See this issue and apply the same to your PR: https://github.com/kubernetes/charts/issues/1785

Eventually, Helm create will do this for you in a future release. See https://github.com/kubernetes/helm/pull/2845

**Comment by danisla** [comment]
@unguiculus, github has a really cool feature that lets you make comments inline. You click on the "Files Changed" tab then the blue "+" icon next to the line number to add a comment. Can you maybe just add a comment with an example of what you are looking for? 

I'll keep diffing my side to see if I can figure it out in the meantime.

File: stable/hadoop/templates/_helpers.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{{/* vim: set filetype=mustache: */}}
+{{/*
+Expand the name of the chart.
+*/}}
+{{- define "name" -}}
```
Comment (unguiculus) [inline stable/hadoop/templates/_helpers.yaml]: We've just started namespacing templates. Please change them like so and apply throughout the chart:

`{{- define "hadoop.name" -}}`
`{{- define "hadoop.fullname" -}}`

File: stable/hadoop/templates/configmaps/hadoop-config.yaml [hunk 2]
```diff
@@ -0,0 +1,296 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ .Release.Name }}-config
```
Comment (unguiculus) [inline stable/hadoop/templates/configmaps/hadoop-config.yaml]: `name: {{ template "hadoop.fullname" . }}`

Only append suffix for components. Fix references below.

File: stable/hadoop/templates/pdbs/hdfs-dn.yaml [hunk 3]
```diff
@@ -0,0 +1,18 @@
+apiVersion: policy/v1beta1
+kind: PodDisruptionBudget
+metadata:
+  name: {{ .Release.Name }}-hdfs-dn
```
Comment (unguiculus) [inline stable/hadoop/templates/pdbs/hdfs-dn.yaml]: `name: {{ template "hadoop.fullname" . }}-hdfs-dn`

File: stable/hadoop/templates/pdbs/hdfs-nn.yaml [hunk 4]
```diff
@@ -0,0 +1,18 @@
+apiVersion: policy/v1beta1
+kind: PodDisruptionBudget
+metadata:
+  name: {{ .Release.Name }}-hdfs-nn
```
Comment (unguiculus) [inline stable/hadoop/templates/pdbs/hdfs-nn.yaml]: `name: {{ template "hadoop.fullname" . }}-hdfs-nn`

File: stable/hadoop/templates/pdbs/yarn-nm.yaml [hunk 5]
```diff
@@ -0,0 +1,18 @@
+apiVersion: policy/v1beta1
+kind: PodDisruptionBudget
+metadata:
+  name: {{ .Release.Name }}-yarn-nm
```
Comment (unguiculus) [inline stable/hadoop/templates/pdbs/yarn-nm.yaml]: `name: {{ template "hadoop.fullname" . }}-yarn-nm`

File: stable/hadoop/templates/pdbs/yarn-rm.yaml [hunk 6]
```diff
@@ -0,0 +1,18 @@
+apiVersion: policy/v1beta1
+kind: PodDisruptionBudget
+metadata:
+  name: {{ .Release.Name }}-yarn-rm
```
Comment (unguiculus) [inline stable/hadoop/templates/pdbs/yarn-rm.yaml]: `name: {{ template "hadoop.fullname" . }}-yarn-rm`

File: stable/hadoop/templates/services/hdfs-dn.yaml [hunk 7]
```diff
@@ -0,0 +1,23 @@
+# A headless service to create DNS records
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ .Release.Name }}-hdfs-dn
```
Comment (unguiculus) [inline stable/hadoop/templates/services/hdfs-dn.yaml]: `name: {{ template "hadoop.fullname" . }}-hdfs-dn`

File: stable/hadoop/templates/services/hdfs-nn.yaml [hunk 8]
```diff
@@ -0,0 +1,23 @@
+# A headless service to create DNS records
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ .Release.Name }}-hdfs-nn
```
Comment (unguiculus) [inline stable/hadoop/templates/services/hdfs-nn.yaml]: `name: {{ template "hadoop.fullname" . }}-hdfs-nn`

File: stable/hadoop/templates/services/yarn-nm.yaml [hunk 9]
```diff
@@ -0,0 +1,24 @@
+# A headless service to create DNS records
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ .Release.Name }}-yarn-nm
```
Comment (unguiculus) [inline stable/hadoop/templates/services/yarn-nm.yaml]: `name: {{ template "hadoop.fullname" . }}-yarn-nm`

File: stable/hadoop/templates/services/yarn-rm.yaml [hunk 10]
```diff
@@ -0,0 +1,20 @@
+# A headless service to create DNS records
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ .Release.Name }}-yarn-rm
```
Comment (unguiculus) [inline stable/hadoop/templates/services/yarn-rm.yaml]: `name: {{ template "hadoop.fullname" . }}-yarn-rm`

File: stable/hadoop/templates/services/yarn-ui.yaml [hunk 11]
```diff
@@ -0,0 +1,18 @@
+# Service to access the yarn web ui
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ .Release.Name }}-yarn-ui
```
Comment (unguiculus) [inline stable/hadoop/templates/services/yarn-ui.yaml]: `name: {{ template "hadoop.fullname" . }}-yarn-ui`

File: stable/hadoop/templates/statefulsets/hdfs-dn.yaml [hunk 12]
```diff
@@ -0,0 +1,96 @@
+apiVersion: apps/v1beta1
+kind: StatefulSet
+metadata:
+  name: {{ .Release.Name }}-hdfs-dn
```
Comment (unguiculus) [inline stable/hadoop/templates/statefulsets/hdfs-dn.yaml]: `name: {{ template "hadoop.fullname" . }}-hdfs-dn`

File: stable/hadoop/templates/statefulsets/hdfs-nn.yaml [hunk 13]
```diff
@@ -0,0 +1,96 @@
+apiVersion: apps/v1beta1
+kind: StatefulSet
+metadata:
+  name: {{ .Release.Name }}-hdfs-nn
```
Comment (unguiculus) [inline stable/hadoop/templates/statefulsets/hdfs-nn.yaml]: `name: {{ template "hadoop.fullname" . }}-hdfs-nn`

File: stable/hadoop/templates/statefulsets/yarn-nm.yaml [hunk 14]
```diff
@@ -0,0 +1,92 @@
+apiVersion: apps/v1beta1
+kind: StatefulSet
+metadata:
+  name: {{ .Release.Name }}-yarn-nm
```
Comment (unguiculus) [inline stable/hadoop/templates/statefulsets/yarn-nm.yaml]: `name: {{ template "hadoop.fullname" . }}-yarn-nm`

File: stable/hadoop/templates/statefulsets/yarn-rm.yaml [hunk 15]
```diff
@@ -0,0 +1,76 @@
+apiVersion: apps/v1beta1
+kind: StatefulSet
+metadata:
+  name: {{ .Release.Name}}-yarn-rm
```
Comment (unguiculus) [inline stable/hadoop/templates/statefulsets/yarn-rm.yaml]: `name: {{ template "hadoop.fullname" . }}-yarn-rm`

File: stable/hadoop/templates/services/yarn-ui.yaml [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+    app: {{ template "name" . }}
+    chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+    component: yarn-ui
+spec:
+  ports:
+  - port: 8088
+    name: web
+  selector:
```
Comment (unguiculus) [inline stable/hadoop/templates/services/yarn-ui.yaml]: Add `release` label to selector.

**Review by unguiculus — CHANGES_REQUESTED** [review]
Thanks for the update. I made some comments.

There's an effort under way to handle storage classes consistently across all chart. Please apply this to your PR as well. See #1869.

Please add node selectors.

You group values by component which is good. The general practice is to append a suffix to the file name that indicates the resource type instead of putting them in directories by resource types. Please do so. If you want to further group the files, create directories reflecting the structure of your values (i. e. hdfs, yarn).

File: stable/hadoop/templates/_helpers.yaml (hunk 1 above)
Comment (danisla) [reply stable/hadoop/templates/_helpers.yaml]: The output of `helm create` with v2.6.1 still generates helpers without the namespace.

**Comment by danisla** [comment]
@unguiculus, thanks for the in-line comments. I think I have addressed them all.

**Comment by viglesiasce** [comment]
Tested manually and was able to get everything working! Thanks for the PR @danisla. Im especially pumped about including an easy way to run tests against the cluster, scale the cluster, and hook in the Zepplin chart.

A++ on the NOTES.txt.

**Review by viglesiasce — APPROVED** [review]

**Review by unguiculus — APPROVED** [review]

**Comment by danisla** [comment]
@viglesiasce, @unguiculus anything else blocking this from being merged?

**Comment by Alanthur** [comment]
@danisla Hi,you use statefulset pods for the hadoop cluster.Can I use Replica Sets pods for the hadoop cluster?Any help is appreciated.

**Comment by danisla** [comment]
@dahaian the hadoop components require a stable identity to communicate, this is why they are created with a StatefulSet.

**Comment by Alanthur** [comment]
@danisla Thanks.Can I use kubernete service for a stable identity?

**Comment by danisla** [comment]
@dahaian K8S services are like load balancers and spread traffic across multiple endpoints. Services do not provide the stable identity required by the hadoop components.

**Comment by Alanthur** [comment]
@danisla Thanks.Generally speaking,the hadoop cluster require in a intranet.But I found that I have to expose ports in a hadoop docker image to communicate.I worry about that the kubernete pods can not communicate before I expose some ports.I do not know whether I expose enough ports to communicate.Can you have any suggestions?Any help is appreciated.

**Comment by danisla** [comment]
@dahaian this chart is intended to be used with in-cluster applications and services such as [zeppelin](https://github.com/kubernetes/charts/tree/master/stable/zeppelin). I don't have any plans right now to extend it to support external communication.

**Comment by Alanthur** [comment]
@danisla Thanks.The hadoop namenode communicate with datanode for some ports.And hadoop namenode is a pod,the datanode is a pod.If I forgot to expose some ports in a hadoop docker image,they would not communicate with each other.Can you have any suggestions?Any help is appreciated.

**Comment by jiaj12** [comment]
@danisla  Hi! I'm using k8s v1.6.4 and encounter a problem when trying to set the hdfs.dataNode.replica value more than 1. Error shows CrashLoopBackoff. Have you encountered this problem? Or is there any specific limitation to the k8s versoin or for the dn amount?

**Pull request merged**